File: dart-funk/lib/src/token.dart
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';

@immutable
class Token extends Equatable {
  // variable values
  const Token.ident(this.value) : type = TokenType.ident;
  const Token.int(this.value) : type = TokenType.int;

  // keywords
  const Token.let()
      : type = TokenType.let,
        value = 'let';
  const Token.function()
      : type = TokenType.function,
        value = 'fn';
  const Token.true_()
      : type = TokenType.true_,
        value = 'true';
  const Token.false_()
      : type = TokenType.false_,
        value = 'false';
  const Token.if_()
      : type = TokenType.if_,
        value = 'if';
  const Token.else_()
      : type = TokenType.else_,
        value = 'else';
  const Token.return_()
      : type = TokenType.return_,
        value = 'return';
  // end of things
  const Token.illegal()
      : type = TokenType.illegal,
        value = r'🤷 ¯\_(ツ)_/¯ 🤷';
  const Token.eof()
      : type = TokenType.eof,
        value = '🛑';
  // operators
  const Token.assign()
      : type = TokenType.assign,
        value = '=';
  const Token.equal()
      : type = TokenType.eq,
        value = '==';
  const Token.notEqual()
      : type = TokenType.ne,
        value = '!=';
  const Token.plus()
      : type = TokenType.plus,
        value = '+';
  const Token.dash()
      : type = TokenType.dash,
        value = '-';
  const Token.bang()
      : type = TokenType.bang,
        value = '!';
  const Token.asterisk()
      : type = TokenType.asterisk,
        value = '*';
  const Token.slash()
      : type = TokenType.slash,
        value = '/';
  const Token.greaterThan()
      : type = TokenType.gt,
        value = '>';
  const Token.lessThan()
      : type = TokenType.lt,
        value = '<';
  // separators
  const Token.comma()
      : type = TokenType.comma,
        value = ',';
  const Token.semicolon()
      : type = TokenType.semicolon,
        value = ';';
  const Token.lParen()
      : type = TokenType.lParen,
        value = '(';
  const Token.rParen()
      : type = TokenType.rParen,
        value = ')';
  const Token.lSquirly()
      : type = TokenType.lSquirly,
        value = '{';
  const Token.rSquirly()
      : type = TokenType.rSquirly,
        value = '}';

  final String value;
  final TokenType type;

  static Token identity(String val) {
    switch (val) {
      case 'let':
        return const Token.let();
      case 'fn':
        return const Token.function();
      case 'true':
        return const Token.true_();
      case 'false':
        return const Token.false_();
      case 'if':
        return const Token.if_();
      case 'else':
        return const Token.else_();
      case 'return':
        return const Token.return_();
      default:
        return Token.ident(val);
    }
  }

  @override
  String toString() => '<${type.name}> => $value';

  @override
  List<Object?> get props => [type, value];
}

enum TokenType {
  ident,
  int,
  let,
  illegal,
  eof,
  assign,
  plus,
  comma,
  semicolon,
  lParen,
  rParen,
  lSquirly,
  rSquirly,
  function,
  dash,
  bang,
  asterisk,
  slash,
  lt,
  gt,
  true_,
  false_,
  if_,
  else_,
  return_,
  ne,
  eq,
}

File: csharp/Monkey/Token.cs
namespace Monkey;

public abstract record Token
{
    public static Token Illegal(char value) => new Illegal(value);
    public static readonly Token EoF = new EoF();
    
    public static Token Identifier(string value) => new Identifier(value);
    public static Token Integer(int value) => new Integer(value);

    public static readonly Token Assign = new Assign();
    public static readonly Token Plus = new Plus();
    public static readonly Token Dash = new Dash();
    public static readonly Token Bang = new Bang();
    public static readonly Token Asterisk = new Asterisk();
    public static readonly Token ForwardSlash = new ForwardSlash();
    public static readonly Token LessThan = new LessThan();
    public static readonly Token GreaterThan = new GreaterThan();
    public static readonly Token NotEqual = new NotEqual();
    public static readonly Token Equal = new Equal();
    
    public static readonly Token Comma = new Comma();
    public static readonly Token Semicolon = new Semicolon();
    
    public static readonly Token LParen = new LParen();
    public static readonly Token RParen = new RParen();
    public static readonly Token LSquirly = new LSquirly();
    public static readonly Token RSquirly = new RSquirly();
    
    public static readonly Token Function = new Function();   
    public static readonly Token Let = new Let();
    public static readonly Token True = new True();
    public static readonly Token False = new False();
    public static readonly Token If = new If();
    public static readonly Token Else = new Else();
    public static readonly Token Return = new Return();
};

public record Illegal(char Value) : Token;
public record EoF : Token;
public record Identifier(string Value): Token;
public record Integer(int Value) : Token;
public record Assign: Token;
public record Plus: Token;

public record Dash : Token;
public record Bang: Token;
public record Asterisk: Token;
public record ForwardSlash: Token;
public record LessThan: Token;
public record GreaterThan: Token;
public record NotEqual: Token;
public record Equal: Token;
public record Comma: Token;
public record Semicolon: Token;
public record LParen: Token;
public record RParen: Token;
public record LSquirly: Token;
public record RSquirly: Token;
public record Function: Token;
public record Let : Token;
public record True: Token;
public record False: Token;
public record If: Token;
public record Else: Token;
public record Return: Token;
